File: CookieClickerSwift/CookieClickerSwift/model/GameState.swift
//
//  GameState.swift
//  CookieClickerSwift
//
//  Created by Alperen Saraç on 14.09.2025.
//

import Foundation

struct GameState: Codable {
    var score: Double = 0
    var cps: Double = 0
    var baseTap: Int = 1
    var extraTap: Int = 0
    var prestigeLevel: Int = 0
    var prestigePoints: Int = 0
    // İstersen perk seviyeleri eklenir
}

final class Save {
    private static let key = "cookie_state_v1"
    static func load() -> GameState {
        guard let data = UserDefaults.standard.data(forKey: key),
              let s = try? JSONDecoder().decode(GameState.self, from: data) else { return GameState() }
        return s
    }
    static func save(_ s: GameState) {
        if let data = try? JSONEncoder().encode(s) {
            UserDefaults.standard.set(data, forKey: key)
        }
    }
}

File: flutter_cookie_clicker/lib/model/game_state.dart
class GameState {
  final double score;
  final double cps;
  final int baseTap;
  final int extraTap;
  final int prestigeLevel;

  const GameState({
    this.score = 0.0,
    this.cps = 0.0,
    this.baseTap = 1,
    this.extraTap = 0,
    this.prestigeLevel = 0,
  });

  GameState copyWith({
    double? score,
    double? cps,
    int? baseTap,
    int? extraTap,
    int? prestigeLevel,
  }) {
    return GameState(
      score: score ?? this.score,
      cps: cps ?? this.cps,
      baseTap: baseTap ?? this.baseTap,
      extraTap: extraTap ?? this.extraTap,
      prestigeLevel: prestigeLevel ?? this.prestigeLevel,
    );
  }

  factory GameState.fromJson(Map<String, dynamic> json) => GameState(
    score: (json['score'] ?? 0.0).toDouble(),
    cps: (json['cps'] ?? 0.0).toDouble(),
    baseTap: json['baseTap'] ?? 1,
    extraTap: json['extraTap'] ?? 0,
    prestigeLevel: json['prestigeLevel'] ?? 0,
  );

  Map<String, dynamic> toJson() => {
    'score': score,
    'cps': cps,
    'baseTap': baseTap,
    'extraTap': extraTap,
    'prestigeLevel': prestigeLevel,
  };
}
